### ISSUE TITLE
[Feature]: Model-aware learned adversarial distortion generator (gradient-based attacks)

### ISSUE THREAD

**Issue by Adit-Jain-srm** [Adit-Jain-srm/NightmareNet#164]
## Problem

The learned adversarial path (`nightmarenet/distortions/learned.py`) uses a generic MLM (DistilBERT) to generate perturbations via attention-weighted token replacement. It's not model-aware - it targets tokens that the MLM considers important, not tokens that the TARGET model being trained is most sensitive to. This means the adversarial examples don't exploit the specific weaknesses of the model under training, limiting the effectiveness of the nightmare phase.

Currently, `LearnedAdversarialGenerator`:
- Uses `_get_token_importance()` based on MLM attention weights (generic, not target-model-specific)
- Replaces important tokens with MLM predictions (plausible but not maximally adversarial for the target)
- Is disabled by default (`learned: 0.0` in config) partly because it's slow (loads a full MLM per call) and not demonstrably better than rule-based distortions

**Relevant files:**
- `nightmarenet/distortions/learned.py` (`LearnedAdversarialGenerator` class - current implementation)
- `nightmarenet/distortions/nightmare.py` (calls learned distortion when `learned > 0.0`)
- `nightmarenet/distortions/registry.py` (plugin registration - new strategies register here)
- `configs/default.yaml` (`distortion.adversarial.learned: 0.0`, `learned_model: "distilbert-base-uncased"`)

## Proposed Solution

Make the learned adversarial generator model-aware by using the TARGET model's gradients (not a separate MLM's attention) to identify vulnerable tokens:

1. **Gradient-based importance scoring** - Replace the current attention-based `_get_token_importance()` with gradient-based importance: compute `grad(loss, embeddings)` on the target model, use the L2 norm of per-token embedding gradients as importance scores. Tokens where small changes in embeddings cause large loss changes are the most vulnerable.

2. **Gradient-guided substitution** - For the top-k important tokens, use the embedding gradient direction to find substitutions that INCREASE loss (not just plausible MLM replacements). Project gradient onto the embedding matrix to find the nearest token in the adversarial direction.

3. **Caching** - Cache generated adversarial examples per cycle. As the model improves across cycles, previously adversarial examples may no longer be effective, so regenerate each cycle but keep the cache for curriculum scheduling (replay earlier adversarial examples alongside new ones).

4. **Config extension:**
   ```yaml
   distortion:
     adversarial:
       learned: 0.3
       learned_model: "distilbert-base-uncased"  # fallback MLM when target model unavailable
       learned_strategy: "gradient"  # "attention" (current) | "gradient" (new, model-aware)
       learned_cache: true            # cache adversarial examples across cycles
   ```

## Acceptance Criteria

- [ ] `LearnedAdversarialGenerator` accepts an optional `target_model` parameter for gradient-based importance
- [ ] Gradient-based importance: `grad(loss, embeddings).norm(dim=-1)` used to rank tokens
- [ ] Gradient-guided substitution: project gradient onto embedding matrix for adversarial token selection
- [ ] Falls back to current attention-based method when `target_model` not provided (backward compatible)
- [ ] Config: `learned_strategy: "gradient"` selects new method, `"attention"` preserves current behavior
- [ ] Adversarial example caching across cycles (configurable)
- [ ] Benchmark: compare gradient-based vs attention-based on robustness improvement for same compute budget
- [ ] Test: gradient-based produces different (ideally more adversarial) outputs than attention-based for same input
- [ ] Test: deterministic under fixed seed
- [ ] No regression when `learned: 0.0` (disabled path unchanged)
- [ ] Performance: gradient-based should not be >3x slower than attention-based per sample

## Technical Notes

- **Gradient computation requires `torch.enable_grad()` context** since the nightmare phase may run under `torch.no_grad()` for evaluation. Ensure the gradient computation is isolated.
- **Memory:** Computing gradients on the full model for each sample is expensive. Use `torch.autograd.grad` with `create_graph=False` to avoid storing the computation graph.
- **The target model is available in `NightmarePhase`** - it's `self.model`. The challenge is passing it through to the distortion generator, which currently operates at the data-generation level (before training). Consider: generate adversarial examples at the START of each nightmare epoch using the current model weights, not during data preparation.

## Scope

Large. ~6-8 hours. Requires understanding of PyTorch autograd, embedding spaces, and the existing distortion pipeline. Research-level implementation.

---

> **Important:** Please read [CONTRIBUTING.md](https://github.com/Adit-Jain-srm/NightmareNet/blob/main/CONTRIBUTING.md) in full before requesting assignment. Assignment requests without a technical approach will not be considered. PRs demonstrating thorough research, benchmarks, or creative approaches beyond minimum criteria may earn bonus XP labels.

**Comment by Jidnyasa-P** [comment]
@Adit-Jain-srm I’d like to take this issue and implement it in a model-aware way.

Plan: update "nightmarenet/distortions/learned.py" so "LearnedAdversarialGenerator" can accept an optional "target_model" and a "learned_strategy" flag. When "learned_strategy="gradient"", I’ll compute token importance from "grad(loss, embeddings).norm(dim=-1)" on the target model, then use the gradient direction to choose substitutions that maximize loss rather than relying only on MLM attention. I’ll keep the current attention-based path as a fallback when "target_model" is not provided, so the existing behavior stays backward compatible.

I’ll also add a small cache layer for adversarial examples across cycles, wire the new config options into "configs/default.yaml", and update "nightmarenet/distortions/nightmare.py" / "nightmarenet/distortions/registry.py" if needed so the new strategy is selectable cleanly from the existing distortion pipeline.

Files I expect to touch: "nightmarenet/distortions/learned.py", "nightmarenet/distortions/nightmare.py", "nightmarenet/distortions/registry.py", and "configs/default.yaml", plus tests for determinism, fallback behavior, and comparison against the current attention-based method.

Estimated timeline: about 1–2 days, since this needs PyTorch autograd work, careful compatibility handling, and benchmarking under the same compute budget.

Kindly assign me this issue under ECSoC'26.

**Comment by Adit-Jain-srm** [comment]
Assigned, go ahead @Jidnyasa-P
